File: mqttclients/dotnet/MQTTnet.Client.Extensions/TelemetryProducer.cs

using MQTTnet.Protocol;

namespace MQTTnet.Client.Extensions;

public class TelemetryProducer<T>
{
    private readonly IMqttClient _mqttClient;
    private readonly string _telemetryTopic;
    private readonly IMessageSerializer _serializer;

    public TelemetryProducer(IMqttClient mqttClient, IMessageSerializer serializer, string topicPattern)
    {
        _mqttClient = mqttClient;
        _serializer = serializer;
        _telemetryTopic = topicPattern!.Replace("{clientId}", _mqttClient.Options.ClientId);
    }

    public Task<MqttClientPublishResult> SendTelemetryAsync(
        T message,
        CancellationToken ct = default) =>
            SendTelemetryAsync(message, MqttQualityOfServiceLevel.AtLeastOnce, false, ct);

    public Task<MqttClientPublishResult> SendTelemetryAsync(
        T message,
        MqttQualityOfServiceLevel qos = MqttQualityOfServiceLevel.AtLeastOnce,
        bool retain = false,
        CancellationToken ct = default) =>
            _mqttClient.PublishBinaryAsync(
                _telemetryTopic,
                _serializer.ToBytes(message),
                qos,
                retain,
                ct);
}

File: mqttclients/ts/mqttjsClientExtensions/src/telemetryProducer.ts
import {
    IPublishPacket,
    MqttClient
} from 'mqtt';
import {
    QoS
} from 'mqtt-packet';
import { IMessageSerializer } from './';

export class TelemetryProducer<T>
{
    private mqttClient: MqttClient;
    private telemetryTopic: string;
    private serializer: IMessageSerializer;

    constructor(mqttClient: MqttClient, serializer: IMessageSerializer, topicPattern: string) {
        this.mqttClient = mqttClient;
        this.serializer = serializer;
        this.telemetryTopic = topicPattern.replace('{clientId}', mqttClient.options.clientId!);
    }

    public async SendTelemetryAsync(message: T, qos: QoS = 1, retain = false): Promise<IPublishPacket> {
        const pubAck = await this.mqttClient.publishAsync(this.telemetryTopic, this.serializer.toBytes(message), {
            qos,
            retain
        });

        return pubAck as IPublishPacket;
    }
}
